src/rover/main.cpp

```diff
@@ -0,0 +1,174 @@
+	if((net_err != network::Error::OK)) {
+		std::cerr << "[!]Failed to connect to rover!" << std::endl;
+	}
+
+	/*
+	 * MAIN LOOP
+	 * 1. Process Info Locally
+	 * 2. Send out packets
+	 */
+	while(true) {
+		/*
+		 * 1. Process Info Locally
+		 * - Grabbing frames
+		 */
+#ifdef TIME
+	auto start_all = high_resolution_clock::now();
+#endif
+		//Grab a frame
+		uint8_t* frame_buffer;
+		std::size_t frame_size;
+		err1 = camera::grab_frame(&session0, &frame_buffer, &frame_size);
+		frame_counter ++;
+#ifdef DEBUG
+		// Print camera buffer info
+		std::cout << (void*)frame_buffer << std::endl;
+		std::cout << "1" << camera::get_error_string(err1) << std::endl;
+#endif
+#ifdef TIME
+	auto end_grab_frame = high_resolution_clock::now();
+	std::cout << 
+		"[LOCAL_TIME] Time to GRAB FRAME from session0: " <<
+		duration_cast<milliseconds>(end_grab_frame - start_all).count() <<
+		"ms" <<
+	std::endl;
+#endif
+
+
+
+		/*
+		 * 2. Send out packets
+		 * - Send Logging packets
+		 * - Create camera messages and send Camera packets
+		 */
+		network::poll_incoming(&conn);
+		network::Message message;
+		// Receive incoming messages //
+		while(network::dequeue_incoming(&conn, &message)) {
+				switch(message.type) {
+					case network::MessageType::HEARTBEAT:
+					{
+						network::Buffer* outgoing = network::get_outgoing_buffer();
+						network::queue_outgoing(&conn, network::MessageType::HEARTBEAT, outgoing);
+						break;
+					}
+					default:
+						break;
+				}
+		}
+		// THIS IS SAMPLE CODE to send outgoing packets //
+		// Log messages
+		network::Buffer* log_buffer = network::get_outgoing_buffer();
+		char s[] = "[LOG] Hewwo Wowld!";
+		network::LogMessage log_message = {static_cast<uint8_t>(strlen(s)), s};
+		network::serialize(log_buffer, &log_message);
+
+		network::queue_outgoing(&conn, network::MessageType::LOG, log_buffer);
+
+		// Camera Frames
+		//Temporary 'for' loop to use the same frame from session0 for all streams
+		for (unsigned int i = 0; i < 4; i++) {
+			uint8_t num_buffers = (frame_size/CAMERA_MESSAGE_SIZE) + 1;
+			for (unsigned int j = 0; j < num_buffers; j++) {
+				network::Buffer* camera_buffer = network::get_outgoing_buffer();
+				// If we are not at the last buffer
+				uint16_t buffer_size = (j != num_buffers-1)?CAMERA_MESSAGE_SIZE:frame_size%CAMERA_MESSAGE_SIZE;
+				network::CameraMessage message = {
+					static_cast<uint8_t>(i),              // stream_index
+					static_cast<uint16_t>(frame_counter), // frame_index
+					static_cast<uint8_t>(j),              // section_index
+					num_buffers,                          // section_count
+					buffer_size,                          // size
+					nullptr                               // data
+				};
+				message.data = (uint8_t*) malloc(CAMERA_MESSAGE_SIZE);
+				// OPTIMIZE: reuse the same message object to avoid reallocating memory
+				memcpy(message.data, frame_buffer+CAMERA_MESSAGE_SIZE*j, CAMERA_MESSAGE_SIZE);
```

Comment:
Ordering is a bit hard to get right away
```suggestion
				memcpy(message.data, j*CAMERA_MESSAGE_SIZE+frame_buffer, CAMERA_MESSAGE_SIZE);
```

Reply:
I've added spaces and parens to make the distinction clear